Feature: Allow Seeding Skills from JSON

ardelato opened the pull request:
## Description

This updates the `DefaultSkills` seeder to read skills from a JSON file located at config/skills.json. If the file does not exist or contains invalid JSON, it falls back to a predefined set of hardcoded skills.

This retains the original functionality of the seeder, but also allows us to easily modify the skills data without having to modify the codebase. This will be useful for our ArgoCD deployments.

Additionally, this adds a new env var to determine if we should truncate the DB and add the skills, or merge the skills into the DB. We would want to truncate the DB in local dev or during testing but in production we would want to ensure that the skills that were manually added are not removed.

qa_req 0

ardelato committed:
feat(skills): load skills from JSON configuration or fallback to hardcoded data

Updated the DefaultSkills seeder to read skills from a JSON file located at
config/skills.json. If the file does not exist or contains invalid JSON, it falls
back to a predefined set of hardcoded skills.

This retains the original functionality of the seeder, but also allows us
to easily modify the skills data without having to modify the codebase.
This will be useful for our ArgoCD deployments.

ardelato committed:
feat(skills): add skills.json to .gitignore

We don't need to commit the skills.json file as it is a configuration
file that is not needed in the repository.

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -14,21 +14,47 @@ public function run(): void
+            $data = [];
+            foreach ($raw as $categoryId => $skills) {
+                foreach ($skills as $skill) {
+                    $entry = [
+                        'skill_name' => $skill['skill_name'],
+                        'category' => (int)$categoryId,
+                    ];
+                    if (isset($skill['description'])) {
+                        $entry['description'] = $skill['description'];
+                    }
+                    $data[] = $entry;
```
[minor]

```suggestion
                    $data[] = [
                        'skill_name' => $skill['skill_name'],
                        'category' => (int)$categoryId,
                        'description' => $skill['description'] ?? null; // Or '' if that's the default.
                    ];
```

ardelato replied:
Addressed in [4755298](https://github.com/iFixit/restarters/pull/17/commits/475529806862e5f13aa8876e695536c8b1b673c0)

mlahargou reviewed (approved):
CR 📱

ardelato commented:
dev_block ⚡ Going to add some functionality surrounding the truncation

ardelato committed:
feat(seeding): add SEEDING_TRUNCATE_SKILLS env variable for skill seeding

Introduced a new environment variable SEEDING_TRUNCATE_SKILLS to control
whether the skills table should be truncated before seeding.

This will be useful for our ArgoCD deployments where we don't want to
truncate the skills table every time we deploy, as it would remove manually
added skills.

ardelato committed:
refactor(seeding): simplify skill entry creation in DefaultSkills seeder

Refactored the DefaultSkills seeder to simplify the creation of skill entries
data array.

Co-authored-by: Michael Lahargou <michaellahargou@gmail.com>

ardelato committed:
refactor(seeding): extract code blocks into separate methods

Updated the DefaultSkills seeder to extract code blocks into separate methods
for better readability and maintainability.

ardelato commented:
un_dev_block ⚡

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -12,50 +13,105 @@ class DefaultSkills extends Seeder
+            $added = 0;
+            
+            foreach ($data as $entry) {
+                if (!DB::table('skills')->where('skill_name', $entry['skill_name'])->exists()) {
```
Can different categories have the same `skill_name`? Do we want this to check categoryid and skill_name?

mlahargou replied:
Also, if the categoryid and skill_name are unchanged, but we want to update the description, this won't do that. Not sure if there's a use case for that.

ardelato replied:
Addressed in [999901e](https://github.com/iFixit/restarters/pull/17/commits/999901e73a15493e104ce2aa20ad2ffdcabb4ebe)

ardelato replied:
We will be able to update the skill description now, but we still only check the existence of a skill based on the name alone.

The only case where I think we'd have a skill in two different categories is if we had first placed it in the wrong category and then tried to correct it by placing it in the other. Otherwise, it does not make sense to have a skill in two different categories:

![SnagitHelper2023 2025-05-13 10 39 53](https://github.com/user-attachments/assets/d1eeee4d-269d-4dcc-ac72-329879288240)

In which case, we'd manually delete the skill from the wrong category.

ardelato commented:
deploy_block ⚡ on suggestion

ardelato committed:
refactor(skills): catch case where we want to update skills description

In the case where we add a description to a skill in our JSON file, we
would not be able to update the record in the database. This is because
we only check if the skill exists in the database.

Let's add an additional check to see if the description is empty and we
want to update the description. In this case, we'll update the record.

ardelato commented:
un_deloy_block ⚡

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+        foreach ($data as $entry) {
+            $skill = DB::table('skills')->where('skill_name', $entry['skill_name'])->first();
+
+            if (is_null($skill)) {
```
[minor]
```suggestion
            if (!$skill) {
```

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+                continue;
+            }
+
+            if (empty($skill->description) && !empty($entry['description'])) {
```
```suggestion
            if ($skill->description != $entry['description'] ?? '') {
```

ardelato replied:
This has the opposite affect of what I was going for.

This will make the JSON file entry always take precedence. Meaning whatever is set in the JSON file will overwrite the DB entry, including dropping the description.

The original idea was to only modify the DB entry if its empty and we actually have a description we want to update it with. That way if we manually update the skill, it would not be overwritten.

mlahargou replied:
Okay, then you can simplify it with: `!$skill->description && $entry['description'] ?? null`

mlahargou replied:
I was thinking that if we ever called the script with updates to the description, we would want to update the value in the DB. But sounds like that's not the case.

ardelato replied:
Addressed in [103387a](https://github.com/iFixit/restarters/pull/17/commits/103387aa27045586d59f984aa491d5d7d0c44650)

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+        $modified = 0;
+
+        foreach ($data as $entry) {
+            $skill = DB::table('skills')->where('skill_name', $entry['skill_name'])->first();
```
Trying to get my brain wrapped around laravel db objects. Since there's a `Skill` Model, can't this be:

```
$skill = Skill::where('skill_name', $entry['skill_name'])->first();
```

or maybe:

```
$skill = Skill::firstWhere('skill_name', $entry['skill_name']);
```

mlahargou replied:
It also looks like there is a `firstOrCreate` method that may be handy. I don't know about here because it will be hard to keep track of the `$added` value.

ardelato replied:
The `$skill` variable is just an array. I thought about using the `Skills` model which is an [Eloquent ORM](https://laravel.com/docs/12.x/eloquent), but thought its not worth it here. This is a script that would get ran as part of the init process and not really part of the core app, so it didn't make since to integrate the `App/Models` namespace here.

mlahargou replied:
How is the `$skill` var an array if you are accessing properties? i.e. `$skill->description`?

In php, you can't access array values with the arrow notation.

<img width="770" alt="image" src="https://github.com/user-attachments/assets/cd961b03-b2be-47fa-8bf4-ab80ffad6106" />

ardelato replied:
Ah my bad! That's correct, it's an `stdClass` object not an array.

That said, the thought behind not using the `Skills` model still stand.

mlahargou replied:
Sounds good. I'm fine with that 👍

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+            }
+
+            if (empty($skill->description) && !empty($entry['description'])) {
+                DB::table('skills')->where('id', $skill->id)->update(['description' => $entry['description']]);
```
If we switch to `Skill::firstWhere`, we can do something like this which will read easier:

```
$skill->description = $entry['description'];
$skill->save()
```

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+
+        if ($truncate) {
+            Log::info('Truncating and re-inserting all skills');
+            DB::table('skills')->truncate();
+            DB::table('skills')->insert($data);
```
```suggestion
            Skills::truncate();
            Skills::insert($data);
```

mlahargou commented on database/seeders/DefaultSkills.php:
```diff
@@ -29,7 +86,42 @@ public function run(): void
+            $skill = DB::table('skills')->where('skill_name', $entry['skill_name'])->first();
+
+            if (is_null($skill)) {
+                DB::table('skills')->insert($entry);
```
Use Eloquent model: `Skills::insert($entry);`

ardelato replied:
Reference: https://github.com/iFixit/restarters/pull/17#discussion_r2089677601

mlahargou reviewed (approved):
CR 📱 I think it works as is. Some minor improvements if you want.

ardelato committed:
refactor(seeder): simplify null check for skills seeder

Co-authored-by: Michael Lahargou <michaellahargou@gmail.com>

ardelato committed:
refactor(seeder): simplify skills description check condition

Co-authored-by: Michael Lahargou <michaellahargou@gmail.com>

ardelato committed:
fix(docker): add config/skills.json to dockerignore

Even though the file is not included in git, it might be included when
we build the Docker image.

mlahargou reviewed (approved):
CR 📱
